src/Provider/PhpBenchUsageProvider.php

```diff
@@ -0,0 +1,287 @@
+
+use Composer\InstalledVersions;
+use PhpBench\Attributes\AfterMethods;
+use PhpBench\Attributes\BeforeMethods;
+use PhpBench\Attributes\ParamProviders;
+use PhpParser\Node;
+use PHPStan\Analyser\Scope;
+use PHPStan\BetterReflection\Reflection\Adapter\ReflectionMethod;
+use PHPStan\Node\InClassNode;
+use PHPStan\PhpDocParser\Lexer\Lexer;
+use PHPStan\PhpDocParser\Parser\PhpDocParser;
+use PHPStan\PhpDocParser\Parser\TokenIterator;
+use ShipMonk\PHPStan\DeadCode\Graph\ClassMethodRef;
+use ShipMonk\PHPStan\DeadCode\Graph\ClassMethodUsage;
+use ShipMonk\PHPStan\DeadCode\Graph\UsageOrigin;
+use function array_map;
+use function array_merge;
+use function explode;
+use function in_array;
+use function is_array;
+use function is_string;
+use function sprintf;
+use function strpos;
+use function substr;
+use function trim;
+
+final class PhpBenchUsageProvider implements MemberUsageProvider
+{
+
+    private bool $enabled;
+
+    private PhpDocParser $phpDocParser;
+
+    private Lexer $lexer;
+
+    public function __construct(
+        ?bool $enabled,
+        PhpDocParser $phpDocParser,
+        Lexer $lexer
+    )
+    {
+        $this->enabled = $enabled ?? InstalledVersions::isInstalled('phpbench/phpbench');
+        $this->phpDocParser = $phpDocParser;
+        $this->lexer = $lexer;
+    }
+
+    public function getUsages(
+        Node $node,
+        Scope $scope
+    ): array
+    {
+        if (!$this->enabled || !$node instanceof InClassNode) { // @phpstan-ignore phpstanApi.instanceofAssumption
+            return [];
+        }
+
+        $classReflection = $node->getClassReflection();
+        $className = $classReflection->getName();
+
+        if (substr($className, -5) !== 'Bench') {
+            return [];
+        }
+
+        $usages = [];
+
+        foreach ($classReflection->getNativeReflection()->getMethods() as $method) {
+            $methodName = $method->getName();
+
+            $paramProviderMethods = array_merge(
+                $this->getParamProvidersFromAnnotations($method->getDocComment()),
+                $this->getParamProvidersFromAttributes($method),
+            );
+
+            foreach ($paramProviderMethods as $paramProvider) {
+                $usages[] = $this->createUsage(
+                    $className,
+                    $paramProvider,
+                    sprintf('Param provider method, used by %s', $methodName),
+                );
+            }
+
+            if ($this->isBenchmarkMethod($method)) {
+                $usages[] = $this->createUsage($className, $methodName, 'Benchmark method');
+            }
+
+            if (!$this->isBeforeOrAfterMethod($method, $className)) {
+                continue;
+            }
+
+            $usages[] = $this->createUsage($className, $methodName, 'Before/After method');
+        }
+
+        return $usages;
+    }
+
+    private function isBenchmarkMethod(ReflectionMethod $method): bool
+    {
+        return strpos($method->getName(), 'bench') === 0;
```

Comment:
Missing check for public visibility?

Reply:
phpbench picks up even private methods 🤷🏾‍♀️